File: my-folder/0143-reorder-list/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reorderList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: None Do not return anything, modify head in-place instead.
        """
        #use slow-fast pointers to know the start point of the array
        slow=head
        fast=head.next
        while(fast and fast.next):
            #to check fast is not null and we didn't reach the end of the array
            slow=slow.next
            fast=fast.next.next
        #second stores the elements in reverse order from the secondhalf
        second=slow.next
        slow.next=prev=None
        while second:
            tmp=second.next
            second.next=prev
            prev=second
            second=tmp
        
        first,second=head,prev
        while second:
            tmp1=first.next
            tmp2=second.next
            first.next=second
            second.next=tmp1
            first=tmp1
            second=tmp2

        #creating two variable of none to store the previous link details and also updating slow.next to None 


        

File: my-folder/0143-reorder-list/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) return;

       ListNode slow = head;
       ListNode fast = head;
       while(fast != null && fast.next != null){
        slow = slow.next;
        fast = fast.next.next;
       }

       ListNode second = reverseList(slow.next);
       slow.next = null; // break the list after that node 

       ListNode first = head;
       while(second != null){
        ListNode temp1 = first.next;
        ListNode temp2 = second.next;

        first.next = second;
        second.next = temp1;

        first = temp1;
        second = temp2;
       }
    }
    
    private ListNode reverseList(ListNode node){
        ListNode prev = null;
        while(node != null) {
            ListNode temp = node.next;
            node.next = prev;
            prev = node;
            node = temp;
        }

        return prev;
    }
}
//      s
// 1 -> 2 -> 3 -> 4  null
//           f
//           s
// 1 -> 2 -> 3 -> 4 -> 5
//                     f
// Input: head = [1,2,3,4]
// 1 -> 2   null <- 3 <- 4
// h1                    h2
// get length of the list and when the mid value is reached, break the list from the half and reverse the second half 
// iterate the second half and reverse the nodes in second half
// then do this
// create dummynode 
// create curr = dummynode // copy node for traversal and list building 
// while(h2 != null)
// dummynode.next = h1
// h1 = h1.next
// dummynode = dummynode.next
// dummynode.next = h2 
// h2 = h2.next
// dummynode = dummynode.next

// while h1 != null
// dummynode.next = h1

// return curr.next and return the list
// Output: [1,4,2,3]
